File: compiler/src/visit/definitions.ts
import { Fact } from "../db";
import type { Node } from "../node";
import type { VariablePatternNode } from "../nodes/patterns/variable";
import type { ConstantDefinitionNode } from "../nodes/statements/constant-definition";
import type { InstanceDefinitionNode } from "../nodes/statements/instance-definition";
import type { TraitDefinitionNode } from "../nodes/statements/trait-definition";
import type { TypeDefinitionNode } from "../nodes/statements/type-definition";
import type { TypeParameterNode } from "../nodes/types/parameter";
import type {
    ConstantAttributes,
    InstanceAttributes,
    TraitAttributes,
    TypeAttributes,
} from "./attributes";

export class Defined extends Fact<Definition> {
    display(): string {
        return "is a definition";
    }
}

export abstract class Definition<N extends Node = Node> {
    node: N;
    comments: string[];

    constructor(node: N, comments: string[]) {
        this.node = node;
        this.comments = comments;
    }
}

export class VariableDefinition extends Definition<VariablePatternNode> {
    value: Node;

    constructor(node: VariablePatternNode, value: Node) {
        super(node, []);
        this.value = value;
    }
}

export class ConstantDefinition extends Definition<ConstantDefinitionNode> {
    attributes: ConstantAttributes;
    value: { assigned: true; node: Node } | { assigned: false; type: Node };

    constructor(
        node: ConstantDefinitionNode,
        comments: string[],
        attributes: ConstantAttributes,
        type: Node,
    ) {
        super(node, comments);
        this.attributes = attributes;
        this.value = { assigned: false, type };
    }
}

export class TypeDefinition extends Definition<TypeDefinitionNode> {
    attributes: TypeAttributes;
    parameters: TypeParameterNode[];

    constructor(
        node: TypeDefinitionNode,
        comments: string[],
        attributes: TypeAttributes,
        parameters: TypeParameterNode[],
    ) {
        super(node, comments);
        this.attributes = attributes;
        this.parameters = parameters;
    }
}

export class TraitDefinition extends Definition<TraitDefinitionNode> {
    attributes: TraitAttributes;
    parameters: TypeParameterNode[];

    constructor(
        node: TraitDefinitionNode,
        comments: string[],
        attributes: TraitAttributes,
        parameters: TypeParameterNode[],
    ) {
        super(node, comments);
        this.attributes = attributes;
        this.parameters = parameters;
    }
}

export class InstanceDefinition extends Definition<InstanceDefinitionNode> {
    attributes: InstanceAttributes;
    value: Node | undefined;

    constructor(
        node: InstanceDefinitionNode,
        comments: string[],
        attributes: InstanceAttributes,
        value: Node | undefined,
    ) {
        super(node, comments);
        this.attributes = attributes;
        this.value = value;
    }
}

export class TypeParameterDefinition extends Definition<TypeParameterNode> {
    constructor(node: TypeParameterNode) {
        super(node, []);
    }
}

export class MarkerConstructorDefinition extends Definition<Node> {
    constructor(node: Node, comments: string[]) {
        super(node, comments);
    }
}

export class StructureConstructorDefinition extends Definition<Node> {
    fields: Map<string, Node>;

    constructor(node: Node, comments: string[], fields: Map<string, Node>) {
        super(node, comments);
        this.fields = fields;
    }
}

export class VariantConstructorDefinition extends Definition<Node> {
    index: number;

    constructor(node: Node, index: number) {
        super(node, []);
        this.index = index;
    }
}

File: compiler/visit/definitions.go
package visit

import (
	"wipple/database"
	"wipple/nodes/attributes"
)

type DefinedFact struct {
	Definition Definition
}

func (fact DefinedFact) String() string {
	return "is a definition"
}

type Definition interface {
	GetNode() database.Node
	GetComments() []string
}

type VariableDefinition struct {
	Node  database.Node
	Value database.Node
}

func (definition *VariableDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *VariableDefinition) GetComments() []string {
	return nil
}

type ConstantDefinition struct {
	Node       database.Node
	Comments   []string
	Attributes ConstantAttributes
	Assigned   bool
	Value      database.Node
}

func (definition *ConstantDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *ConstantDefinition) GetComments() []string {
	return definition.Comments
}

type ConstantAttributes struct {
	Unit bool
}

func ParseConstantAttributes(attrs []*attributes.AttributeNode) ConstantAttributes {
	return ConstantAttributes{
		Unit: attributes.HasNameAttribute(attrs, "unit"),
	}
}

type TypeDefinition struct {
	Node       database.Node
	Comments   []string
	Attributes TypeAttributes
	Parameters []database.Node
}

func (definition *TypeDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *TypeDefinition) GetComments() []string {
	return definition.Comments
}

type TypeAttributes struct {
	Intrinsic bool
}

func ParseTypeAttributes(attrs []*attributes.AttributeNode) TypeAttributes {
	return TypeAttributes{
		Intrinsic: attributes.HasNameAttribute(attrs, "intrinsic"),
	}
}

type TraitDefinition struct {
	Node       database.Node
	Comments   []string
	Attributes TraitAttributes
	Parameters []database.Node
}

func (definition *TraitDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *TraitDefinition) GetComments() []string {
	return definition.Comments
}

type TraitAttributes struct{}

func ParseTraitAttributes(attrs []*attributes.AttributeNode) TraitAttributes {
	return TraitAttributes{}
}

type InstanceDefinition struct {
	Node       database.Node
	Comments   []string
	Attributes InstanceAttributes
	Value      database.Node
	Trait      database.Node
}

func (definition *InstanceDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *InstanceDefinition) GetComments() []string {
	return definition.Comments
}

type InstanceAttributes struct {
	Default bool
	Error   bool
}

func ParseInstanceAttributes(attrs []*attributes.AttributeNode) InstanceAttributes {
	return InstanceAttributes{
		Default: attributes.HasNameAttribute(attrs, "default"),
		Error:   attributes.HasNameAttribute(attrs, "error"),
	}
}

type TypeParameterDefinition struct {
	Node database.Node
}

func (definition *TypeParameterDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *TypeParameterDefinition) GetComments() []string {
	return nil
}

type MarkerConstructorDefinition struct {
	Node     database.Node
	Comments []string
}

func (definition *MarkerConstructorDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *MarkerConstructorDefinition) GetComments() []string {
	return definition.Comments
}

type StructureConstructorDefinition struct {
	Node     database.Node
	Comments []string
	Fields   map[string]database.Node
}

func (definition *StructureConstructorDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *StructureConstructorDefinition) GetComments() []string {
	return definition.Comments
}

type VariantConstructorDefinition struct {
	Node  database.Node
	Index int
}

func (definition *VariantConstructorDefinition) GetNode() database.Node {
	return definition.Node
}

func (definition *VariantConstructorDefinition) GetComments() []string {
	return nil
}
